amplipi/ctrl.py

```diff
@@ -608,8 +625,168 @@ def get_stations(self, id, stream_index=None):
+      return g['id'] + 1
+    else:
+      return 10000
+
+  @utils.save_on_success
+  def create_preset(self, preset):
+    try:
+      # Make a new preset and add it to presets
+      # TODO: validate preset
+      id = self._new_preset_id()
+      preset['id'] = id
+      self.status['presets'].append(preset)
+      return preset
+    except Exception as e:
+      return utils.error('create preset failed: {}'.format(e))
+
+  @utils.save_on_success
+  def set_preset(self, id, preset_changes):
+    i, preset = utils.find(self.status['presets'], id)
+    configurable_fields = ['name', 'commands', 'state']
+    if i is None:
+      return utils.error('Unable to find preset to redefine')
+
+    try:
+      # TODO: validate preset
+      for f in configurable_fields:
+        if f in preset_changes:
+          preset[f] = preset_changes[f]
+    except Exception as e:
+      return utils.error('Unable to reconfigure preset {}: {}'.format(id, e))
+
+  @utils.save_on_success
+  def delete_preset(self, id):
+    """Deletes an existing preset"""
+    try:
+      i, _ = utils.find(self.status['presets'], id)
+      if i is not None:
+        del self.status['presets'][i] # delete the cached preset state just in case
+    except KeyError:
+      return utils.error('delete preset failed: {} does not exist'.format(id))
+
+
+  def _effected_zones(self, preset_id):
+    """ Aggregate the zones that will be modified by changes """
+    st = self.status
+    _, preset = utils.find(st['presets'], preset_id)
+    effected = set()
+    if preset is None:
+      return effected
+    ps = preset['state']
+    src_zones = { src['id'] : [ z['id'] for z in st['zones'] if z['source_id'] == src['id']] for src in st['sources'] }
+    if 'sources' in ps:
+      for src in ps['sources']:
+        if 'id' in src:
+          effected.update(src_zones[src['id']])
+    if 'groups' in ps:
+      for g in ps['groups']:
+        if 'id' in g:
+          _, gf = utils.find(st['groups'], g['id'])
+          effected.update(gf['zones'])
+    if 'zones' in ps:
+      for z in ps['zones']:
+        if 'id' in z:
+          effected.add(z['id'])
+    return effected
+
+  @utils.save_on_success
+  def load_preset(self, id):
+    """
+    To avoid any issues with audio coming out of the wrong speakers, we will need to carefully load a preset configuration. Below is an idea of how a preset configuration could be loaded to avoid any weirdness. We are also considering adding a "Last config" preset that allows us to easily revert the configuration changes.
+
+    1. Grab system modification mutex to avoid accidental changes (requests during this time return some error)
+    1. Save current configuration as "Last config" preset
+    1. Mute any effected zones
+    1. Execute changes to source, zone, group each in increasing order
+    1. Unmute effected zones that were not muted
+    1. Execute any stream commands
+    1. Release system mutex, future requests are successful after this
```

Comment:
Is this supposed to be a list of steps? Why are they all '1.' ?

Reply:
These were copied from a markdown document, should probably number them so they are easier to read